packages/toolpad-app/src/toolpad/AppEditor/PagesExplorer/index.tsx

```diff
@@ -166,22 +193,73 @@ export default function PagesExplorer({ className }: PagesExplorerProps) {
     if (appDom.isPage(node)) {
       appStateApi.setView({ kind: 'page', nodeId: node.id });
     }
+  };
 
-    if (appDom.isCodeComponent(node)) {
-      appStateApi.setView({ kind: 'codeComponent', nodeId: node.id });
-    }
+  const handlerTreeRef = React.useRef<HTMLUListElement>(null);
+
+  React.useEffect(() => {
+    handlerTreeRef.current?.querySelector(`.${treeItemClasses.selected}`)?.scrollIntoView();
+  }, []);
+
+  const [newPageInput, setNewPageInput] = React.useState('');
+
+  const {
+    value: isCreateNewPageOpen,
+    setTrue: handleOpenCreateNewPage,
+    setFalse: handleCloseCreateNewPageDialog,
+  } = useBoolean(false);
+
+  const nextProposedName = React.useMemo(
+    () => appDom.proposeName(DEFAULT_NEW_PAGE_NAME, existingNames),
+    [existingNames],
+  );
 
-    if (appDom.isConnection(node)) {
-      appStateApi.setView({ kind: 'connection', nodeId: node.id });
+  const handleCreateNewPage = React.useCallback(() => {
+    setNewPageInput(nextProposedName);
+    handleOpenCreateNewPage();
+  }, [handleOpenCreateNewPage, nextProposedName]);
+
+  const handleCloseCreateNewPage = React.useCallback(() => {
+    setNewPageInput('');
+    handleCloseCreateNewPageDialog();
+  }, [handleCloseCreateNewPageDialog]);
+
+  const [anchorEl, setAnchorEl] = React.useState<HTMLButtonElement | null>(null);
+  const createNewInputRef = React.useRef(null);
+  const open = !!anchorEl;
+
+  const newPageInputErrorMsg = useNodeNameValidation(newPageInput, existingNames, 'page');
+
+  React.useEffect(() => {
+    setAnchorEl(newPageInputErrorMsg ? createNewInputRef.current : null);
+  }, [newPageInputErrorMsg]);
+
+  const handleCreateNewCommit = React.useCallback(async () => {
+    if (!newPageInput || newPageInputErrorMsg) {
+      handleCloseCreateNewPage();
+      return;
     }
-  };
 
-  const [createPageDialogOpen, setCreatePageDialogOpen] = React.useState(0);
-  const handleCreatePageDialogOpen = React.useCallback((event: React.MouseEvent) => {
-    event.stopPropagation();
-    setCreatePageDialogOpen(Math.random());
-  }, []);
-  const handleCreatepageDialogClose = React.useCallback(() => setCreatePageDialogOpen(0), []);
+    const newNode = appDom.createNode(dom, 'page', {
+      name: newPageInput,
+      attributes: {
+        title: newPageInput,
+        display: 'shell',
+      },
+    });
+    const appNode = appDom.getApp(dom);
+
+    appStateApi.update((draft) => appDom.addNode(draft, newNode, appNode, 'pages'), {
+      kind: 'page',
+      nodeId: newNode.id,
+    });
+
+    handleCloseCreateNewPage();
+
+    setTimeout(() => {
```

Comment:
`setTimeout` doesn't feel 100% right to me. I mean, I don't think there is a strong guarantee that the item has rendered. If it needs to run after a render, shouldn't it need to run in an effect?

Reply:
This usually always works in my experience as I just want it to wait for the next render. I did it with effects but in both cases I needed to use state just for this.